Enable Zipkin tracing

fabolive opened the pull request:
This PR makes Envoy use the code added by PR #692 (Zipkin library).

This PR replaces the old PR #700.

cc: @mattklein123, @htuch, @RomanDzhabarov

fabolive commented:
@moderation Fixed!

ccaraman commented on source/common/tracing/zipkin_tracer_impl.h:
```diff
@@ -0,0 +1,197 @@
+   * annotation then this Envoy instance initiated the span. Only the Envoy that initiates a Zipkin
+   * span should add binary annotations to it; otherwise, duplicate binary annotations
+   * would be created.
+
```
nit: missing *

fabolive replied:
Done

ccaraman commented on source/common/tracing/zipkin_tracer_impl.h:
```diff
@@ -0,0 +1,197 @@
+
+   * Note that Tracing::HttpTracerUtility::finalizeSpan() makes several calls to this function,
+   * associating several key-value pairs with this span.
+   */
```
nit: please document the params.

ccaraman commented on source/server/configuration_impl.cc:
```diff
@@ -109,27 +110,42 @@ void MainImpl::initializeTracers(const Json::Object& configuration) {
+    http_tracer_.reset(
+        new Tracing::HttpTracerImpl(std::move(lightstep_driver), server_.localInfo()));
+  } else if (type == "zipkin") {
+    if (server_.localInfo().clusterName().empty()) {
```
nit: remove else statement and add ASSERT checking for type is zipkin.  ex https://github.com/lyft/envoy/blob/master/source/common/router/router_ratelimit.cc#L102

fabolive replied:
Done

goaway commented on source/common/json/config_schemas.cc:
```diff
@@ -954,6 +954,26 @@ const std::string Json::Schema::TOP_LEVEL_CONFIG_SCHEMA(R"EOF(
+        "properties" : {
+          "type" : {
+            "type" : "string",
+            "enum" : ["zipkin"]
```
I might have missed some prior discussion, but my reading of this enum is that tracing drivers were originally envisioned to all be specified under a single "driver" schema, with a type ("zipkin", "lightstep") specified via this field.

Most of the other fields are identical across "lightstep_driver" and "zipkin_driver."

fabolive replied:
My original code had both "lightstep" and "zipkin" as valid values of the same enum, and I had made  `access_token_file` an optional parameter, as it has no use for Zipkin. However, under PR #700 (the old PR replaced by this one), @RomanDzhabarov suggested that the following: "this should be done in a way that there are separate schemas for different driver types, otherwise schema will not have much value."

fabolive commented:
@goaway I addressed two of your comments above. Please, see my reply to your comment on driver schema.

RomanDzhabarov commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+SpanPtr ZipkinDriver::startSpan(Http::HeaderMap& request_headers, const std::string&,
+                                SystemTime start_time) {
+  Zipkin::Tracer& tracer = *tls_.getTyped<TlsZipkinTracer>(tls_slot_).tracer_;
+  ZipkinSpanPtr active_span;
```
can you move this declaration to the very bottom, close to the usage.

RomanDzhabarov commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+
+    context.populateFromString(request_headers.OtSpanContext()->value().c_str());
+
+    // Create either a child or a shared-context Zipkin span.
```
can you put a comment on when a child is created vs when shared-context Zipkin span.

RomanDzhabarov commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+  }
+
+  // Set the sampled header.
+  request_headers.insertXB3Sampled().value(std::string(("1")));
```
can you create static const string with the name ALWAYS_SAMPLE or something instead of std::string(("1"))

RomanDzhabarov commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+void ZipkinReporter::onSuccess(Http::MessagePtr&& http_response) {
+  if (Http::Utility::getResponseStatus(http_response->headers()) !=
+      enumToInt(Http::Code::Accepted)) {
+    driver_.tracerStats().reports_dropped_.inc();
```
do you want to differentiate this failure from general onFailure? on line 181

RomanDzhabarov commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+
+#include "spdlog/spdlog.h"
+
+namespace Tracing {
```
can you run coverage test to make sure that coverage for new code is 100%

RomanDzhabarov commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+#include "common/tracing/http_tracer_impl.h"
+#include "common/tracing/zipkin/zipkin_core_constants.h"
+
+#include "spdlog/spdlog.h"
```
is this used? can you do sanity checks on included headers

fabolive commented on source/common/tracing/zipkin/tracer.h:
```diff
@@ -78,18 +86,40 @@ class Tracer : public TracerInterface {
+   * @return the reporter associated with the Tracer.
+   * This is only implemented to support the move constructor.
+   */
+  ReporterPtr reporter() { return std::move(reporter_); }
```
If I do that the code does not pass asan/tsan checks. The following error is reported:
`moving a temporary object prevents copy elision [-Werror,-Wpessimizing-move]`

goaway commented:
We're going to collide here https://github.com/lyft/envoy/pull/932, but I don't think it's going to be that bad. If you address Roman's comments, we may be able to get you in first.

mattklein123 commented:
@fabolive I haven't looked at this since it's not passing tests yet. Will be happy to help move this along once it is, so let me know.

fabolive commented:
@mattklein123, @goaway. I should commit my changes soon.

fabolive commented:
@mattklein123, @goaway: I have addressed @RomanDzhabarov's comments.

RomanDzhabarov commented:
Thanks @fabolive i'll look more

fabolive commented:
@RomanDzhabarov FYI: I will add 2 more (minor) tests soon that are needed.

RomanDzhabarov commented:
got it, i also see https://s3.amazonaws.com/lyft-envoy/coverage/report-master/coverage.source_common_tracing_zipkin_zipkin_core_types.cc.html
can you improve coverage here?

fabolive commented:
@RomanDzhabarov: 

I added the tests I was planning to in zipkin_tracer_impl_test (100%) coverage. 

Also, regarding your latest comments, I have addressed them as follows: (1) As the previously-uncovered Span assignment operator was actually not being used at all, I removed the code; (2) I added a test to cover the creation of a shared-context span with a parent in the preceding context.

fabolive commented:
@mattklein123 With my latest commit I have completed all work I planned to do as part of this PR. If there are no further comments, this work is finished.

mattklein123 commented:
I will review first thing tomorrow. Thanks @fabolive very excited to have this land.

mattklein123 commented on docs/configuration/overview/tracing.rst:
```diff
@@ -43,3 +44,26 @@ access_token_file
+collector_endpoint
+  *(required, string)* The API endpoint of the Zipkin service where the
+  spans will be sent. When using a standard Zipkin installation, the
+  API endpoint is typically `/api/v1/spans`.
```
Is it worth making /api/v1/spans the default if not set and documenting that?

mattklein123 commented on source/common/tracing/zipkin/span_buffer.cc:
```diff
@@ -2,7 +2,7 @@
 namespace Zipkin {
 
-bool SpanBuffer::addSpan(SpanPtr&& span) {
+bool SpanBuffer::addSpan(Span&& span) {
```
Why are we moving away from passing pointers here? I'm fairly certain that if you define a copy constructor on Span (which you have) and you don't define an explicit move constructor, you are actually going to be copying. This is much less efficient. You also have to run destructors in the buffer array every time you clear. I strongly suspect that you want to revert these changes, but not totally sure why they were made. Can potentially help with design if there is more context.

fabolive replied:
I forgot to define the Span move constructor!  It was my intention to have defined the move constructor.  Defining it will avoid the copies.

The reason I moved away from passing pointers above is as follows. Envoy deals with a unique_ptr to a span. That is enforced by the the virtual method Tracing::Driver::startSpan(), which ZipkinDriver and LightStepDriver implement.
Envoy also defines the virtual method Span::finishSpan(), which must be implemented by LightStepSpan and ZipkinSpan. This method has to perform the needed actions on the span before it can be sent to the corresponding backend (LightStep or Zipkin). One of the things that both the LightStep and the Zipkin-related code do when "finishing" a span is to buffer it so that a batch of spans can be sent together to the backend.  Thus, the finish() method of the Span class must insert "this" span into the buffer. It turns out that doing so while passing around a unique_ptr is not safe. Thus, both the LightStep and the Zipkin code do a std::move(*this) to pass the object via move semantics into the buffer (Granted, I forgot to define the Span move constructor!).  It seems to me that we can either pass the object via move semantics (as it is done today), or change the Envoy base Tracing interface to, perhaps, use a shared_ptr to a Span instead. The latter will require changes to the Lightstep-related code.

mattklein123 replied:
OK, be aware though that you are going to run destructors multiple times even if you implement move constructor. I'm OK with this for now as long as the move constructor is correctly implemented and there is a comment somewhere as to why it is necessary. We may want to change the interfaces to use shared_ptr and shared_from_this but I wouldn't do that now. We can see what shows up on perf traces.

mattklein123 replied:
To make it clear that we are currently copying here, please make this method take a ```const Span&``` instead. I would also add a TODO that this needs to be looked at for perf.

mattklein123 commented on source/common/tracing/zipkin/tracer.h:
```diff
@@ -44,23 +44,33 @@ typedef std::unique_ptr<Reporter> ReporterPtr;
 class Tracer : public TracerInterface {
 public:
   /**
+   * Move constructor.
```
Like my other comment, I would avoid using the move constructor here. You aren't actually moving most things, and it makes the code harder to follow. I would remove the move constructor, and refactor the code so that if you have to move a tracer, you should actually be moving a unique_ptr to a tracer.

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -71,6 +71,13 @@ Annotation& Annotation::operator=(const Annotation& ann) {
 
+void Annotation::changeEndpointServiceName(const std::string& service_name) {
+  if (endpoint_.valid()) {
+    Endpoint& endpoint = const_cast<Endpoint&>(endpoint_.value());
```
using const_cast is generally an indication that there is a design problem. Please refactor so const_cast is not necessary. If you need design help we can discuss the specifics.

mattklein123 commented on source/server/configuration_impl.cc:
```diff
@@ -109,27 +110,41 @@ void MainImpl::initializeTracers(const Json::Object& configuration) {
+        new Tracing::HttpTracerImpl(std::move(lightstep_driver), server_.localInfo()));
+  } else {
+    ASSERT(type == "zipkin");
+    if (server_.localInfo().clusterName().empty()) {
```
Can we move this check up above the driver stuff so we don't copy it. You can just say that cluster name must be defined if tracing is enabled.

mattklein123 commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,201 @@
+#include "common/tracing/zipkin_tracer_impl.h"
```
Can we move this file into the zipkin directory and the zipkin namespace. There really isn't any "library" anymore so I see no reason to have it be outside that namespace.

fabolive replied:
In the very initial discussions we had with @RomanDzhabarov, he suggested that the Zipkin-related code, similar to the LightStep code, should be under the Tracing namespace. In fact, lightstep_tracer_impl.cc and zipkin_tracer_impl.cc are quite similar. I see those files as some sort of conceptual template for the implementation of other tracing drivers.
Do you feel strongly about moving that to the zipkin directory and to the Zipkin namespace?

mattklein123 replied:
The LS code should really be in a LS namespace. There is no reason for this code to be outside the namespace of all the other Zipkin code. Please move it.

mattklein123 commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,201 @@
+                           Runtime::RandomGenerator& random_generator)
+    : cm_(cluster_manager),
+      tracer_stats_{ZIPKIN_TRACER_STATS(POOL_COUNTER_PREFIX(stats, "tracing.zipkin."))}, tls_(tls),
+      runtime_(runtime), local_info_(local_info), tls_slot_(tls.allocateSlot()) {
```
tls_slot_ can be const

mattklein123 commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,201 @@
+  }
+  cluster_ = cluster->info();
+
+  if (cluster_->features() & Upstream::ClusterInfo::Features::HTTP2) {
```
I don't think this is really true. There is no reason that a zipkin collector cannot be behind an http/2 proxy. I would remove this check.

rshriram replied:
Thats a fair point. I put this in there, coz the typical use case is a zipkin server (Java based server) and it doesn't seem to have HTTP/2. I guess, there could be a Envoy there as well.

mattklein123 commented on source/common/tracing/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,201 @@
+
+ZipkinReporter::ZipkinReporter(ZipkinDriver& driver, Event::Dispatcher& dispatcher,
+                               const std::string& collector_endpoint)
+    : driver_(driver), collector_endpoint_(collector_endpoint) {
```
collector_endpoint_ can be const

fabolive commented:
@mattklein123 Please, se my new commits addressing your comments.

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -71,6 +71,14 @@ Annotation& Annotation::operator=(const Annotation& ann) {
 
+void Annotation::changeEndpointServiceName(const std::string& service_name) {
+  if (endpoint_.valid()) {
+    Endpoint endpoint(endpoint_.value());
```
You can avoid copying here by just adding a non-const value() function here: https://github.com/lyft/envoy/blob/master/include/envoy/common/optional.h#L37

The method signature should be: ```T& value()``` and the body will be the same as the other one.

mattklein123 replied:
Sorry just to clarify, when you add the new non-const version, have the const version call the non-const version to avoid code duplication. :)

fabolive replied:
I cannot have one call the other without using `const_cast`. Are you okay is this use of `const_cast`?

mattklein123 replied:
The const version can call the non-cost version. Basically never use const_cast.

fabolive replied:
That seems to be intuitive, indeed. That's what I did. 
The problem here is different, though. Irrespective of whether or not  `const_cast` is used, the compiler seems to detect an infinite recursion. Perhaps because of optimization? (I am compiling the code with the CI docker image, the one I pulled today).
So, the compiler is not happy about both versions below:
*Version1:*
```
T& value() {
    if (!valid_) {
      throw EnvoyException("fetching invalid Optional value");
    }
    return value_;
}
const T& value() const { return value(); }
```
*Version 2:*
```
T& value() {
    if (!valid_) {
      throw EnvoyException("fetching invalid Optional value");
    }
    return value_;
}
const T& value() const { return const_cast<const T&>(value()); }
```

mattklein123 replied:
I see, sorry, yes. I'm not sure off the top of my head the right way to deal with this. I would just copy the code and then make sure you have test coverage for both variants in optional_test (just declare a const Optional for testing).

fabolive replied:
ok. I will duplicate the code. I believe avoiding code duplication requires an ugly combination of `const_cast` and `static_cast`...

mattklein123 commented on source/common/tracing/zipkin/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+  return false;
+}
+
+ZipkinDriver::TlsZipkinTracer::TlsZipkinTracer(TracerPtr tracer, ZipkinDriver& driver)
```
nit: for clarify making this take ```TracerPtr&&``` which will make it clear at the call site that you are moving.

mattklein123 commented on source/common/tracing/zipkin/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,194 @@
+
+namespace Zipkin {
+
+ZipkinSpan::ZipkinSpan(Zipkin::Span& span) : span_(span) {}
```
nit: it would be optimal if you remove Zipkin from all of the class names now, since you are inside the Zipkin namespace. You can veto this but I would appreciate it. 😉

fabolive replied:
I only left `Zipkin::` in the `ZipkinSpan` class.  It turns out that, because `ZipkinSpan` inherits from `Tracing::Span`, any reference to `Span` is considered to be a reference to `Tracing::Span` and not to `Zipkin::Span`.  I explicitly refer to `Zipkin::Span` to disambiguate. I was surprised by this behavior. Any suggestion?

mattklein123 replied:
No that's fine.

mattklein123 commented on test/common/tracing/zipkin/tracer_test.cc:
```diff
@@ -26,31 +27,36 @@ class TestReporterImpl : public Reporter {
-  Tracer tracer("my_service_name", addr);
-  MockMonotonicTimeSource mock_start_time;
-  MonotonicTime start_time = mock_start_time.currentTime();
+  Runtime::RandomGeneratorImpl random_generator;
```
Use MockRandomGenerator

fabolive replied:
I can certainly do that; no problem. The reason I used the real RandomGenerator in some cases was that `MockRandomGenerator::random()` always returns 0, therefore it renders some of my checks useless. It would be nice if we could control the value to be returned by `MockRandomGenerator::random()` to some extent, e.g., zero or non-zero.

mattklein123 replied:
You can. It's a mock, have it return whatever you want in your tests.

fabolive replied:
I am now using Mock, but had to comment out a few checks. I left the checks commented out (instead of removing them) because they document expected code behavior and perhaps we wight  be able to un-comment them one day. I can remove them if you prefer.

mattklein123 replied:
Yes, please remove commented out code. The point of the mock is so that you can return whatever value want, and can exercise all of your code paths completely and reproducibly. So make sure you have full coverage when using all mocks, and remove all commented out checks.

mattklein123 commented on test/common/tracing/zipkin/zipkin_core_types_test.cc:
```diff
@@ -461,6 +506,59 @@ TEST(ZipkinCoreTypesSpanTest, copyConstructor) {
+
+  Span span2(std::move(span));
+
+  EXPECT_EQ(id, span2.id());
```
If you are going to bother writing a move constructor test, you should sanity check that span1 has actually been moved from, and a copy was not done.

fabolive replied:
According to the test coverage reports with and without my implementation of the move constructor, this test is indeed exercising the move constructor.
Given that a "move" in the pure sense of the word does not happen when `std::move` is invoked, I don't think there is a safe/good way to determine whether or not an object was moved from. One (ugly) approach to automatically verify if the move constructor was actually called  would be to change the class implementation to set a member flag to a value in the case of a copy constructor execution, and to another value when the move constructor is exercised. I don't think we want that.

mattklein123 replied:
You can check span1 to see if it still contains data members. The move should literally move them, so string should be empty, etc.

fabolive replied:
I don't think that is the case. My point is: there is no move in the literal sense. I believe what actually happens is some sort of "optimized copy". In fact, the data members of the original span seem to be intact.

mattklein123 replied:
Then you aren't actually moving anything. In the interest of moving this forward, I would kill the move constructor and just do the copy, and deal with the perf issues in a follow up.

mattklein123 commented on test/common/tracing/zipkin/zipkin_tracer_impl_test.cc:
```diff
@@ -0,0 +1,315 @@
+
+  void setupValidDriver() {
+    EXPECT_CALL(cm_, get("fake_cluster")).WillRepeatedly(Return(&cm_.thread_local_cluster_));
+    ON_CALL(*cm_.thread_local_cluster_.cluster_.info_, features())
```
IIRC this is the default so this line can be removed.

mattklein123 commented on test/common/tracing/zipkin/zipkin_tracer_impl_test.cc:
```diff
@@ -0,0 +1,315 @@
+  NiceMock<Runtime::MockLoader> runtime_;
+  NiceMock<ThreadLocal::MockInstance> tls_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  Runtime::RandomGeneratorImpl random_;
```
Mock

mattklein123 replied:
(Just search all the test code for non-mock versions to make sure you get them all). Thanks.

mattklein123 commented:
P.S., @rshriram @fabolive I would strongly recommend setting up a perf run with zipkin enabled and getting some (perf) traces. I think there is going to be some super low hanging fruit. Can obviously be done in a follow up.

fabolive commented:
@mattklein123 Please, see my latest commit and my replies to your comments. Do you have any suggestions based on all my comments above?

mattklein123 commented:
@fabolive yes, I responded.

mattklein123 commented on source/common/tracing/zipkin/zipkin_core_types.cc:
```diff
@@ -73,9 +73,8 @@ Annotation& Annotation::operator=(const Annotation& ann) {
 void Annotation::changeEndpointServiceName(const std::string& service_name) {
   if (endpoint_.valid()) {
-    Endpoint endpoint(endpoint_.value());
+    Endpoint& endpoint = endpoint_.value();
```
nit: delete this line and just do ```endpoint_.value().setServiceName(service_name)```

fabolive commented:
@mattklein123 Let me know if there are any further changes that are needed.

mattklein123 commented:
@fabolive I still see most of the classes have Zipkin in them in the Zipkin namespace. Are we skipping that for now? (I thought your comment meant you had changed all of them besides for ZipkinSpan).

fabolive commented:
@mattklein123 I actually meant something else. I might have misunderstood your original comment. Never mind; let me rename the classes.

fabolive commented:
@mattklein123 Classes renamed (except for ZipkinSpan).

mattklein123 commented on source/common/tracing/zipkin/tracer.h:
```diff
@@ -15,12 +15,12 @@ namespace Zipkin {
  * span will be sent to out to Zipkin, or buffered so that it can be sent out later.
  */
-class Reporter {
+class ReporterInterface {
```
To match the rest of the code, please leave this as Reporter, and make the implementation class ReporterImpl

mattklein123 commented on source/common/tracing/zipkin/zipkin_tracer_impl.cc:
```diff
@@ -0,0 +1,193 @@
+#include "common/tracing/http_tracer_impl.h"
+#include "common/tracing/zipkin/zipkin_core_constants.h"
+
+#include "spdlog/spdlog.h"
```
I think @RomanDzhabarov already mentioned this, but this include should not be needed here.

fabolive replied:
If I remove that include statement the code does not compile because of this:
```
throw EnvoyException(fmt::format("{} collector cluster is not defined on cluster manager level",
                                     config.getString("collector_cluster")));
```
As you can see, I am actually using `fmt::format()`. Do you prefer that I do not use `fmt::format` in favor of plain string concatenation.?

mattklein123 replied:
Ah OK, that's fine. (Didn't seeing logging so was confused). No problem.

fabolive commented:
@mattklein123 Code updated.

rshriram commented:
@kyessenov time to update Envoy SHA in our code
